v0/destinations/sendgrid/util.js

```diff
@@ -361,12 +413,185 @@ const generatePayloadFromConfig = (payload, Config) => {
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+        meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_PARAM
+      })
+      .build();
+  }
+};
+
+/**
+ * Returns address in string format
+ * @param {*} address
+ * @returns
+ */
+const flattenAddress = address => {
+  const addressType = typeof address;
+  let companyAddress = "";
+  if (addressType === "object") {
+    const keys = Object.keys(address);
+    keys.forEach(key => {
+      companyAddress += `${address[key]} `;
+    });
+  } else {
+    companyAddress = address;
+  }
+  return companyAddress.trim();
+};
+
+/**
+ * Returns contactListIds in string format
+ * contactListIds : it's a string created from an array
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const getContactListIds = (message, destination) => {
+  const { Config } = destination;
+  const contactListIds = [];
+  const listIds = getDestinationExternalID(message, "listIds");
+
+  if (Config.listId) {
+    contactListIds.push(Config.listId);
+  }
+
+  if (listIds && typeof listIds === "string") {
+    contactListIds.push(listIds);
+  }
+
+  if (listIds && typeof listIds === "object" && listIds.length > 0) {
+    contactListIds.push(...listIds);
+  }
+  return contactListIds.toString();
+};
+
+/**
+ * Returns the list of custom_fields
+ * @param {*} destination
+ * @returns
+ */
+const fetchCustomFields = async destination => {
+  const { apiKey } = destination.Config;
+  return customFieldsCache.get(apiKey, async () => {
+    const requestOptions = {
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${apiKey}`
+      }
+    };
+    const endpoint = "https://api.sendgrid.com/v3/marketing/field_definitions";
+
+    const resonse = await httpGET(endpoint, requestOptions);
+    const processedResponse = processAxiosResponse(resonse);
+    if (isHttpStatusSuccess(processedResponse.status)) {
+      const { custom_fields: customFields } = processedResponse.response;
+      return customFields;
+    }
+    throw new ErrorBuilder()
+      .setMessage(
+        "[SendGrid] :: Error while fetching custom_fields, please try after sometime"
+      )
+      .setStatus(400)
+      .setStatTags({
+        destType: DESTINATION,
+        stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM,
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+        meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_EVENT
+      })
+      .build();
+  });
+};
+
+/**
+ * Returns customField object
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} contactDetails
+ * @returns
+ */
+const getCustomFields = async (message, destination) => {
+  const customFields = {};
+  const traits = getFieldValueFromMessage(message, "traits");
+  const contextTraits = get(message, "context.traits");
+  const payload = { ...traits, ...contextTraits };
```

Comment:
```suggestion
  const payload = getFieldValueFromMessage(message, "traits");
```
No need to choose from both places.

Reply:
@ItsSudip  i kept context.traits since custom_fields are user level

Reply:
Okay